umi_tools/whitelist.py

```diff
@@ -0,0 +1,456 @@
+                    "(starting with 'cell_') %s, %s" (
+                        options.pattern, options.pattern2))
+
+    if options.stdin == sys.stdin:
+        read1s = umi_methods.fastqIterate(U.openFile(options.stdin))
+    else:
+        read1s = umi_methods.fastqIterate(U.openFile(options.stdin.name))
+
+    # set up read extractor
+    ReadExtractor = umi_methods.ExtractFilterAndUpdate(
+        method=options.extract_method,
+        pattern=options.pattern,
+        pattern2=options.pattern2,
+        prime3=options.prime3,
+        extract_cell=extract_cell)
+
+    cell_barcode_counts = collections.Counter()
+
+    n_reads = 0
+    n_cell_barcodes = 0
+
+    # if using the umis method, need to keep a set of umis observed
+    if options.method == "umis":
+        cell_barcode_umis = collections.defaultdict(set)
+
+    if not options.read2_in:
+        for read1 in read1s:
+            n_reads += 1
+            barcode_values = ReadExtractor.getBarcodes(read1)
+            if barcode_values is None:
+                continue
+            else:
+                cell, umi, _, _, _, _, _ = barcode_values
+                if options.method == "umis":
+                    cell_barcode_umis[cell].add(umi)
+                else:
+                    cell_barcode_counts[cell] += 1
+                n_cell_barcodes += 1
+
+            if options.subset_reads:
+                if n_cell_barcodes > options.subset_reads:
+                    break
+    else:
+        read2s = umi_methods.fastqIterate(U.openFile(options.read2_in))
+        for read1, read2 in izip(read1s, read2s):
+            n_reads += 1
+
+            barcode_values = ReadExtractor.getBarcodes(read1, read2)
+            if barcode_values is None:
+                continue
+            else:
+                cell, umi, _, _, _, _, _ = barcode_values
+                if options.method == "umis":
+                    cell_barcode_umis[cell].add(umi)
+                else:
+                    cell_barcode_counts[cell] += 1
+                n_cell_barcodes += 1
+
+            if options.subset_reads:
+                if n_reads > options.subset_reads:
+                    break
+
+    if options.method == "umis":
+        for cell in cell_barcode_umis:
+            cell_barcode_counts[cell] = len(cell_barcode_umis[cell])
+
+    if options.cell_number and options.cell_number > len(cell_barcode_counts):
+        raise ValueError(
+            "--set-cell-barcode option specifies more cell barcodes than the "
+            "number of observed barcodes %i/%i" % (
+                options.cell_number, len(cell_barcode_counts)))
+
+    cell_whitelist, true_to_false_map = umi_methods.getCellWhitelist(
+        cell_barcode_counts,
+        options.cell_number,
+        options.error_correct_threshold,
+        options.plot_prefix)
+
+    # re-make the reads1s iterator
+    read1s = umi_methods.fastqIterate(U.openFile(options.stdin.name))
```

Comment:
Thanks for spotting. This bug is left-over from when the code was used to both generate the whitelist and then filter the fastq so the iterator was remade. This would of course cause a fail when `options.stdin`==`sys.stdin`!